### ISSUE TITLE
Functionality to allow for finetuning

### ISSUE THREAD

**Issue by asoyemi1** [general-molecular-simulations/so3lr#3]
Hello developers, I was wondering if your code has functionality to allow for finetuning on user-defined training sets, and if so can you provide a minimal example?

Thank you!

**Comment by VladCarare** [comment]
Hi @kabylda ! I was wondering if there's any plan to add finetuning capabilities? I'm happy to contribute if needed. I've been trying to hack my way around it but I'm so far unsuccessful.

**Comment by VladCarare** [comment]
This is my hacky script, altered from mlff/mlff/cAPI/mlff_train_so3krates.py.

It seems to work, in that the loss and other metrics go down.

Things I will try to look into:
- Freezing some of the parameters
- Whether it's effective in any way to fine-tune on a train set even though it's at a different level of DFT theory

```
import numpy as np
import jax
import jax.numpy as jnp
import logging
import os
import argparse
import wandb
import json
import portpicker

from pathlib import Path
from typing import Dict
from ase.units import *

from mlff.io import create_directory, bundle_dicts, save_dict, load_params_from_ckpt_dir
from mlff.training import Coach, Optimizer, get_loss_fn, create_train_state
from mlff.data import DataTuple, DataSet
from mlff.cAPI.process_argparse import StoreDictKeyPair
from mlff.nn.stacknet import get_obs_and_force_fn, get_observable_fn, get_energy_force_stress_fn
from mlff.nn import So3krates
from mlff.nn.observable import Energy
from mlff.data import AseDataLoader
from mlff.properties import md17_property_keys

import mlff.properties.property_names as pn

# logging.basicConfig(level=logging.INFO)
port = portpicker.pick_unused_port()
jax.distributed.initialize(f'localhost:{port}', num_processes=1, process_id=0)


def unit_convert_data(x: Dict, table: Dict):
    """
    Convert units in the data dictionary.

    Args:
        x (Dict): The data dictionary.
        table (Dict): A dictionary mapping quantities to conversion factors.

    Returns: The data dictionary with converted quantities.

    """
    for (k, v) in x.items():
        if k in list(table.keys()):
            logging.info('Converted {} to ase default unit.'.format(k))
            x[k] *= table[k]
    return x



# Create the parser
prop_keys =md17_property_keys

# Add the arguments
data_file=None
train_data_file='train.xyz'
valid_data_file='test.xyz'

shift_by = 'atomic_number'

shifts ={1:-13.58722251684341,6:-1027.6147175552076,7:-1482.0104237939017,8:-2039.2226254442378}

ckpt_dir="test"

ckpt_manager_options = None

restart_from_ckpt_dir = "so3lr/so3lr/params/checkpoints/ckpt_5150000/params"


# Model Arguments
r_cut = 4.5 
F = 128
L = 3
H = 4
degrees = [1,2,3,4]

so3krates_layer_kwargs = None
zbl_repulsion = True
geometry_embed_kwargs = None
# Structure arguments
mic = False
# Data Arguments
n_train = None
n_valid = None
n_test = None

epochs = 100
steps = None
# steps', type=int, required=False, help='Number of training steps.')
lr_stop = 1e-6

# Arguments that determine the training parameters
batch_size =None
training_batch_size=None
validation_batch_size=None
units=None

# Training arguments
model_seed=0
data_seed=0
training_seed=0

targets=[pn.energy, pn.force]
inputs = [pn.atomic_type, pn.atomic_position, pn.idx_i, pn.idx_j, pn.node_mask]

lr = 1e-5
lr_decay_plateau = None

lr_decay_exp_default = {'transition_steps': 100_000, 'decay_factor': 0.7}
lr_decay_exp = lr_decay_exp_default
lr_warmup = None

clip_by_global_norm = None

default_loss_weights = {pn.energy: 0.01, pn.force: 0.99, pn.stress: 0.01}
loss_weights = default_loss_weights

eval_every_t = None
use_wandb = False
wandb_initdefault={}

def parse_data_file(x):
    if x is not None:
        return Path(x).absolute().resolve().as_posix()
    else:
        return x

data_file = parse_data_file(data_file)
train_data_file = parse_data_file(train_data_file)
valid_data_file = parse_data_file(valid_data_file)

if data_file is None and (train_data_file is None or valid_data_file is None):
    raise ValueError("Either `--data_file` or (`--train_data_file` + `--valid_data_file`) must be specified.")

if data_file is None:
    data_files = [train_data_file, valid_data_file]
else:
    data_files = [data_file]

shift_by = shift_by
shifts = shifts

if shifts is not None:
    shifts = {int(k): float(v) for (k, v) in shifts.items()}

if ckpt_dir is None:
    ckpt_dir = (Path(os.getcwd()).absolute().resolve() / 'module').as_posix()
else:
    ckpt_dir = (Path(ckpt_dir).absolute().resolve()).as_posix()

restart_from_ckpt_dir = None
if restart_from_ckpt_dir is not None:
    restart_from_ckpt_dir = (Path(restart_from_ckpt_dir).absolute().resolve()).as_posix()
    assert restart_from_ckpt_dir != ckpt_dir

if Path(ckpt_dir).exists():
    raise FileExistsError(f'Checkpoint directory {ckpt_dir} already exists.')

jax_dtype = 'x64'
if jax_dtype == 'x64':
    jax.config.update("jax_enable_x64", True)

eval_every_t = eval_every_t
use_wandb = use_wandb

lr = lr
lr_decay_plateau = lr_decay_plateau
if lr_decay_plateau is not None:
    lr_decay_plateau = {k: float(v) for k, v in lr_decay_plateau.items()}

lr_decay_exp = {'exponential': lr_decay_exp} if lr_decay_exp is not None else lr_decay_exp

if mic:
    inputs += [pn.unit_cell]
    inputs += [pn.cell_offset]

_loss_weights = loss_weights
loss_weights = {k: float(v) for (k, v) in _loss_weights.items() if k in targets}

total_loss_weight = sum([x for x in loss_weights.values()])
effective_loss_weights = {k: v / total_loss_weight for k, v in loss_weights.items()}

if data_file is not None:
    if n_train is None or n_valid is None:
        raise ValueError('If only a single `--data_file` is provided, please specify the number of training'
                            'and validation samples via `--n_train` and `--n_valid`.')

conversion_table = {}
if units is not None:
    for (q, v) in units.items():
        k = prop_keys[q]
        conversion_table[k] = eval(v)

all_data = []
for d in data_files:
    extension = os.path.splitext(d)[1]
    if extension == '.npz':
        data = dict(np.load(d))
    else:
        load_stress = pn.stress in targets
        data_loader = AseDataLoader(d, load_stress=load_stress, neighbors_format='dense')
        data = data_loader.load_all()

    data = unit_convert_data(data, table=conversion_table)
    if pn.stress in targets:
        cell_key = prop_keys[pn.unit_cell]
        stress_key = prop_keys[pn.stress]

        stress = data[stress_key]
        try:
            assert stress.shape[-2:] == (3, 3)
        except AssertionError:
            raise ValueError('Stress tensor must be a matrix with shape (3,3). '
                                'Voigt convention not supported yet.')

        # re-scale stress with cell volume
        cells = data[cell_key]  # shape: (B,3,3)
        cell_volumes = np.abs(np.linalg.det(cells))  # shape: (B)
        data[stress_key] = stress * cell_volumes[:, None, None]
    all_data += [data]

if len(all_data) == 2:
    n_train = len(all_data[0][prop_keys[pn.atomic_position]])
    n_valid = len(all_data[1][prop_keys[pn.atomic_position]])

    data = jax.tree_map(lambda x, y: np.concatenate([x, y]), *all_data)

    data_set = DataSet(data=data, prop_keys=prop_keys)
    data_set.index_split(data_idx_train=list(range(n_train)),
                            data_idx_valid=list(range(n_train, int(n_train+n_valid))),
                            data_idx_test=[],
                            r_cut=r_cut,
                            training=True,
                            mic=mic)
elif len(all_data) == 1:
    data = all_data[0]
    data_set = DataSet(data=data, prop_keys=prop_keys)
    data_set.random_split(n_train=n_train,
                            n_valid=n_valid,
                            n_test=n_test,
                            r_cut=r_cut,
                            training=True,
                            mic=mic,
                            seed=data_seed)
else:
    raise RuntimeError('You should not end up here. Please file an issue :-)')

if shift_by == 'mean':
    data_set.shift_x_by_mean_x(x=pn.energy)
elif shift_by == 'atomic_number':
    data_set.shift_x_by_type(x=pn.energy, shifts=shifts)
elif shift_by == 'lse':
    data_set.shift_x_by_type(x=pn.energy)

d = data_set.get_data_split()

scales = {}
loss_variance_scaling = False
if loss_variance_scaling:
    for t in targets:
        if t == pn.stress:
            scales[prop_keys[t]] = 1 / np.nanvar(d['train'][prop_keys[t]], axis=0)
        elif t == pn.energy:
            scales[prop_keys[t]] = 1 / np.nanvar(d['train'][prop_keys[t]])
        elif t == pn.force:
            force_data_train = d['train'][prop_keys[t]]
            node_msk_train = d['train'][prop_keys[pn.node_mask]]
            print(force_data_train.shape)
            print(node_msk_train.shape)
            scales[prop_keys[t]] = 1 / np.nanvar(force_data_train[node_msk_train])
        else:
            raise NotImplementedError('Loss with variance scaling currently only implemented for loss with '
                                        'energy and/or forces and/or stress.')
else:
    scales = None

n_heads = H

so3krates_layer_kwargs = {'degrees': degrees,
                            'n_heads': n_heads}

if so3krates_layer_kwargs is not None:
    so3krates_layer_kwargs.update(so3krates_layer_kwargs)

if zbl_repulsion:
    print('Running with ZBL repulsion.')

geometry_embed_kwargs = {'degrees': degrees,
                            'mic': mic,
                            'r_cut': r_cut}
if geometry_embed_kwargs is not None:
    geometry_embed_kwargs.update(geometry_embed_kwargs)

obs = [Energy(prop_keys=prop_keys, zbl_repulsion=zbl_repulsion)]
net = So3krates(prop_keys=prop_keys,
                F=F,
                n_layer=L,
                obs=obs,
                geometry_embed_kwargs=geometry_embed_kwargs,
                so3krates_layer_kwargs=so3krates_layer_kwargs)

if pn.force in targets:
    if pn.stress in targets:
        obs_fn = get_energy_force_stress_fn(net)
    else:
        obs_fn = get_obs_and_force_fn(net)
else:
    obs_fn = get_observable_fn(net)

obs_fn = jax.vmap(obs_fn, in_axes=(None, 0))

opt = Optimizer(clip_by_global_norm=clip_by_global_norm)

tx = opt.get(learning_rate=lr)

def autoset_batch_size(u):
    if u < 500:
        return 1
    elif 500 <= u < 1000:
        return 5
    elif 1000 <= u < 10_000:
        return 10
    elif u >= 10_000:
        return 100
if batch_size is not None:
    batch_size = batch_size
else:
    batch_size = autoset_batch_size(n_train)

training_batch_size = batch_size if training_batch_size is None else training_batch_size
validation_batch_size = batch_size if validation_batch_size is None else validation_batch_size

if epochs is None and steps is None:
    assert lr_stop is not None
    if lr_decay_exp is None and lr_decay_plateau is None:
        raise ValueError('No learning rate decay is specified. At the same time neither epochs nor steps is speci'
                            f'fied such that the training is stopped when the learning rate is below {lr_stop}. Thus'
                            f'either specify a learning rate decay using the `--lr_decay_exp` or the'
                            f' `lr_decay_plateau` argument. Alternatively, specify a number of epochs, steps.')
    _epochs = 1_000_000_000
elif epochs is None and steps is not None:
    _epochs = int(steps / (n_train / training_batch_size))
elif epochs is not None and steps is None:
    _epochs = epochs
elif epochs is not None and steps is not None:
    raise ValueError('Only epochs or steps argument can be specified.')
else:
    msg = 'One should not end up here. This is likely due to a bug in the mlff package. Please report to ' \
            'https://github.com/thorben-frank/mlff'
    raise RuntimeError(msg)

coach = Coach(inputs=inputs,
                targets=targets,
                epochs=_epochs,
                training_batch_size=training_batch_size,
                validation_batch_size=validation_batch_size,
                loss_weights=effective_loss_weights,
                ckpt_dir=ckpt_dir,
                data_path=data_file,
                train_data_path=data_file if train_data_file is None else train_data_file,
                valid_data_path=data_file if valid_data_file is None else valid_data_file,
                net_seed=model_seed,
                training_seed=training_seed,
                stop_lr_min=lr_stop)

loss_fn = get_loss_fn(obs_fn=obs_fn,
                        weights=effective_loss_weights,
                        scales=scales,
                        prop_keys=prop_keys)

data_tuple = DataTuple(inputs=inputs,
                        targets=targets,
                        prop_keys=prop_keys)

train_ds = data_tuple(d['train'])
valid_ds = data_tuple(d['valid'])

inputs = jax.tree_map(lambda x: jnp.array(x[0, ...]), train_ds[0])
if restart_from_ckpt_dir is None:
    params = net.init(jax.random.PRNGKey(coach.net_seed), inputs)
else:
    print(f"Restarting training from {restart_from_ckpt_dir}.")
    # params = load_params_from_ckpt_dir(restart_from_ckpt_dir)

    from mlff.mdx.potential.mlff_potential_sparse import load_model_from_workdir

    net, params = load_model_from_workdir(
        workdir='/home/ec2-user/Calculations/dev/so3lr/so3lr/params',
        from_file=True,
        model='so3krates',
        long_range_kwargs=dict(
        cutoff_lr=12.,
        dispersion_energy_cutoff_lr_damping=2.,
        neighborlist_format_lr='ordered_sparse',
        coulomb_kspace_do_ewald=False,
        coulomb_kspace_interp_nodes=4,),
        output_intermediate_quantities=None
    )



train_state, h_train_state = create_train_state(net,
                                                params,
                                                tx,
                                                polyak_step_size=None,
                                                plateau_lr_decay=lr_decay_plateau,
                                                scheduled_lr_decay=lr_decay_exp,
                                                lr_warmup=lr_warmup
                                                )

h_net = net.__dict_repr__()
h_opt = opt.__dict_repr__()
h_coach = coach.__dict_repr__()
h_dataset = data_set.__dict_repr__()
h = bundle_dicts([h_net, h_opt, h_coach, h_dataset, h_train_state])

Path(ckpt_dir).mkdir(parents=True, exist_ok=False)
if data_file is not None:
    data_set.save_splits_to_file(ckpt_dir, 'splits.json')

data_set.save_scales(ckpt_dir, 'scales.json')
# save_dict(path=ckpt_dir, filename='hyperparameters.json', data=h, exists_ok=True)

# if use_wandb:
#     wandb.init(config=h, **args.wandb_init)

coach.run(train_state=train_state,
            train_ds=train_ds,
            valid_ds=valid_ds,
            loss_fn=loss_fn,
            eval_every_t=eval_every_t,
            log_every_t=1,
            ckpt_manager_options=ckpt_manager_options,
            restart_by_nan=True,
            use_wandb=use_wandb)

```
